### DISCUSSION TITLE
DNS Zones

### DISCUSSION THREAD

**Discussion by johnmmcgee** [0xERR0R/blocky#1448] [General]
Hello,

I recently saw this commit: https://github.com/0xERR0R/blocky/commit/9f633f18d03246ebc50a26f268453ee94d60ed25

If I am correct about this, that means that blocky will be able to load up a zone file as such:
```
customDNS:
  zone:  $INCLUDE local.zone
```

If so, that is great!

**Comment by ThinkChaos** [comment]
Yes that will be supported in the next release!  
Shoutout to @BenMcH for both extending custom DNS to support CNAME in #1352 and then following up with the zonefile support in #1360.

Since this is for custom DNS, we support A/AAAA and CNAME records. Implementing more is possible but I don't think it's worth trying to support everything because that's a lot of complexity. If you want arbitrary zones, the recommendation is to setup something link bind and point blocky to it.

**Reply by johnmmcgee** [reply]
Yes this is awesome.  

Will there be anything added to the api to help manage the zone file? add, remove and maybe modify?

what about a zone refresh interval?

**Reply by kwitsch** [reply]
The "zone file" is part of the config and is therefore read during the startup process. 
Currently there is no mechanism to reload the config during runtime.

**Reply by johnmmcgee** [reply]
ah ok.  a zone/refresh api call, much like the list/refresh call, would be pretty great.

https://0xerr0r.github.io/blocky/swagger.html#operation--lists-refresh-post

**Reply by ThinkChaos** [reply]
Yeah partially reloading the config is something that generally needs love. I also thought of allowing to change the log level on the fly. Would help to debug stuff.

**Reply by kwitsch** [reply]
@ThinkChaos I remember that we had a discussion about config hot reloading a long time ago... 🤔

Since then we adapted to context usage so maybe we could do some config-context binding to rebuild internal structs instead of reconfiguring them?

**Reply by ThinkChaos** [reply]
I spent a bit of time thinking about it and wrote a comment in #607 with some ideas for future reference.

I think putting config + derived data (for instance lists URLs + parsed lists) in something we pass down with each request would be a nice alternative design. Though I wouldn't put it in a `context.Context` since that's for "request-scoped values", which even the log is pushing it a bit IMO. But in something that can replace/wrap `model.Request` would be nice, and you also get to use the type system, whereas `Context.Value` returns `any` which needs to be cast+checked everywhere.

The main reason I'm leaning towards rebuilding the resolvers from scratch is it's more in line with the current design of resolvers doing that derivation step themselves and managing all their state.  
Migrating to an architecture where the data derivation happens separately seems like it would need to touch a lot of components. And it might be easy to miss some state, and end up with a mix of data that came from different config values.
Starting from scratch, passing everything down on every request would be a cool design: resolvers would be pure functions of their inputs and we'd get config reloading for free from there.

Also this is subjective, but splitting that data derivation step doesn't sound fun to me, but messing with the server would interest me a lot :smile:

**Reply by BenMcH** [reply]
@ThinkChaos Is there a standard release cadence for blocky? I've just been using the `main` tag but would like to get to a versioned release :sweat_smile:

**Reply by ThinkChaos** [reply]
@0xERR0R handles that :)
AFAIU there's no strict cadence cause dev here is often by spurts depending on motivation!

There's a couple things assigned in [the milestone](https://github.com/0xERR0R/blocky/milestone/24).
I just added #1273 to it, I'll take a look at that tomorrow to add a setting to workaround the iOS bug.

**Reply by 0xERR0R** [reply]
I try to make a release each 2-3 month. "Release" means we just build binaries/docker images and do a github release. There is no code freeze / testing phase. Code on master should be stable enough.

Btw, we can think about other release process, it should be not a "single man show".

**Reply by ThinkChaos** [reply]
I guess ideally we'd automate the process so the CI can do everything when a tag is pushed, and sharing that responsibility would be easier

**Reply by 0xERR0R** [reply]
> I guess ideally we'd automate the process so the CI can do everything when a tag is pushed, and sharing that responsibility would be easier

Actually, setting a tag will start Release build. Manual step is to write release notes, because auto-generated look not readable and to release on github

**Reply by ThinkChaos** [reply]
Ok then a `.md` with a list of what to do would be nice I think, just something to be sure nothing is missed

**Reply by 0xERR0R** [reply]
Yeah, maybe following the https://keepachangelog.com/en/1.1.0/ specification. Need to check if there are any github actions to automate it

**Reply by kwitsch** [reply]
If we keep a change log(or generate a nice one) this can be used by goreleaser. 
https://goreleaser.com/customization/release/#custom-release-notes
